serialization/python_pickle.ksy

```diff
@@ -0,0 +1,588 @@
+meta:
+  id: python_pickle
+  title: Python pickle serialization format
+  application: Python
+  file-extension:
+    - pickle
+    - pkl
+  license: CC0-1.0
+  endian: le
+  xref:
+    justsolve: Pickle
+    wikidata: Q7190889
+doc: |
+  Python Pickle format serializes Python objects to a byte stream, as a sequence
+  of operations to run on the Pickle Virtual Machine.
+
+  The format is mostly implementation defined, there is no formal specification.
+  Pickle data types are closely coupled to the Python object model.
+  Python singletons, and most builtin types (e.g. `None`, `int`,`dict`, `list`)
+  are serialised using dedicated Pickle opcodes.
+  Other builtin types, and all classes  (e.g. `set`, `datetime.datetime`) are
+  serialised by encoding the name of a constructor callable.
+  They are deserialised by importing that constructor, and calling it.
+  So, unpickling an arbitrary pickle, using the Python's stdlib pickle module
+  can cause arbitrary code execution.
+
+  Pickle format has evolved with Python, later protocols add opcodes & types.
+  Later Python releases can pickle to or unpickle from any earlier protocol.
+
+  * Protocol 0: ASCII clean, no explicit version, fields are '\n' terminated.
+  * Protocol 1: Binary, no explicit version, first length prefixed types.
+  * Protocol 2: Python 2.3+. Explicit versioning, more length prefixed types.
+    https://www.python.org/dev/peps/pep-0307/
+  * Protocol 3: Python 3.0+. Dedicated opcodes for `bytes` objects.
+  * Protocol 4: Python 3.4+. Opcodes for 64 bit strings, framing, `set`.
+    https://www.python.org/dev/peps/pep-3154/
+doc-ref: https://github.com/python/cpython/blob/v3.7.3/Lib/pickletools.py
+seq:
+  # TODO is there a way to declare PROTO is optional, but only valid at position 0?
```

Comment:
@moreati,  are you going to fix this and that? I have given you info you have requested.

More precisely, I meant something like
```yaml
seq:
  pos:
    - id: first_op
      type: u1
instances:
  after_or_just:
    value: first_op == 0x80
  main_after:
    pos: 1
    type: pickle_main
    if: after_or_just
  main_just:
    pos: 0
    if: not after_or_just
  main:
    value: "(after_or_just?main_after:main_just)"
types:
   pickle_main:
      ....
```

Reply:
I'm leaning toward not doing this here. I'm starting to think of validating inclusion/placement of `PROTO` and `STOP` as semantic concerns - to be checked/enforced at a layer above the parser, along with other op ordering requirements. Encoding _just_ these two checks here makes the format deceleration harder to understand, without sufficient benefit.

Does that sound okay?

Reply:
It is definitely sounds OK,  but because of another reasons: the opcodes validated via `contents` are not true objects. IMHO the KS spec should have this encoded because it is a spec, it should be self-sufficient in the sense it should be possible to throw away the original spec and rely only on Kaitai one. Also we want assertions functionality (run-time toggleable) and a serializer (this implies a symbolic engine and an SMT-solver in order to automatically resolve the assertions and expressions). And IMHO the constraints on the first and last opcode should be encoded as assertions against objects, not raw bytes, but they are not yet implemented.